typescript/sdk/src/ism/EvmIsmReader.ts

```diff
@@ -60,37 +61,57 @@ export class EvmIsmReader implements IsmReader {
+    super(multiProvider, chain);
   }
 
   async deriveIsmConfig(address: Address): Promise<DerivedIsmConfig> {
-    const ism = IInterchainSecurityModule__factory.connect(
-      address,
-      this.provider,
-    );
-    const moduleType: ModuleType = await ism.moduleType();
-    this.logger.debug('Deriving ISM config', { address, moduleType });
-
-    switch (moduleType) {
-      case ModuleType.UNUSED:
-        throw new Error('UNUSED does not have a corresponding IsmType');
-      case ModuleType.ROUTING:
-        // IsmType is either ROUTING or FALLBACK_ROUTING, but that's determined inside deriveRoutingConfig
-        return this.deriveRoutingConfig(address);
-      case ModuleType.AGGREGATION:
-        return this.deriveAggregationConfig(address);
-      case ModuleType.LEGACY_MULTISIG:
-        throw new Error('LEGACY_MULTISIG is deprecated and not supported');
-      case ModuleType.MERKLE_ROOT_MULTISIG:
-      case ModuleType.MESSAGE_ID_MULTISIG:
-        return this.deriveMultisigConfig(address);
-      case ModuleType.NULL:
-        return this.deriveNullConfig(address);
-      case ModuleType.CCIP_READ:
-        throw new Error('CCIP_READ does not have a corresponding IsmType');
-      default:
-        throw new Error('Unknown ModuleType');
+    let moduleType: ModuleType | undefined = undefined;
+    let derivedIsmConfig: DerivedIsmConfig;
+    try {
+      const ism = IInterchainSecurityModule__factory.connect(
+        address,
+        this.provider,
+      );
+      this.logger.debug('Deriving IsmConfig:', { address });
+
+      // Temporarily turn off SmartProvider logging
+      // Provider errors are expected because deriving will call methods that may not exist in the Bytecode
+      this.setSmartProviderLogLevel('silent');
+      moduleType = await ism.moduleType();
+
+      switch (moduleType) {
+        case ModuleType.UNUSED:
+          throw new Error('UNUSED does not have a corresponding IsmType');
+        case ModuleType.ROUTING:
+          // IsmType is either ROUTING or FALLBACK_ROUTING, but that's determined inside deriveRoutingConfig
+          derivedIsmConfig = await this.deriveRoutingConfig(address);
+          break;
+        case ModuleType.AGGREGATION:
+          derivedIsmConfig = await this.deriveAggregationConfig(address);
+          break;
+        case ModuleType.LEGACY_MULTISIG:
+          throw new Error('LEGACY_MULTISIG is deprecated and not supported');
+        case ModuleType.MERKLE_ROOT_MULTISIG:
+        case ModuleType.MESSAGE_ID_MULTISIG:
+          derivedIsmConfig = await this.deriveMultisigConfig(address);
+          break;
+        case ModuleType.NULL:
+          derivedIsmConfig = await this.deriveNullConfig(address);
+          break;
+        case ModuleType.CCIP_READ:
+          throw new Error('CCIP_READ does not have a corresponding IsmType');
+        default:
+          throw new Error(`Unknown ISM ModuleType: ${moduleType}`);
+      }
+    } catch (e: any) {
```

Comment:
I think i prefer using `finally` because the error is being rethrown. In the case that the caller decides to do something else with the error related to the provider, the logging will be silent.

Reply:
good callout– will update for hook reader as well